galley/pkg/source/kube/builtin/source.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builtin
+
+import (
+	"fmt"
+	"istio.io/istio/galley/pkg/runtime"
+	"reflect"
+	"sync"
+
+	"istio.io/istio/galley/pkg/runtime/resource"
+	"istio.io/istio/galley/pkg/source/kube/log"
+	"istio.io/istio/galley/pkg/source/kube/schema"
+	"istio.io/istio/galley/pkg/source/kube/stats"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+// NewSource creates a new built-in source. If the type is not built-in, returns an error.
+func NewSource(sharedInformers informers.SharedInformerFactory, spec schema.ResourceSpec) (runtime.Source, error) {
+	p := processors[spec.Kind]
+	if p == nil {
+		return nil, fmt.Errorf("unknown resource type: name='%s', gv='%v'",
+			spec.Singular, spec.GroupVersion())
+	}
+	return newSource(sharedInformers, p), nil
+}
+
+func newSource(sharedInformers informers.SharedInformerFactory, p *Processor) runtime.Source {
+	return &source{
+		Processor:       p,
+		sharedInformers: sharedInformers,
+	}
+}
+
+type source struct {
+	*Processor
+
+	// Lock for changing the running state of the source
+	stateLock sync.Mutex
+
+	sharedInformers informers.SharedInformerFactory
+	informer        cache.SharedIndexInformer
+
+	// stopCh is used to quiesce the background activity during shutdown
+	stopCh  chan struct{}
+	handler resource.EventHandler
+}
+
+// Start the source. This will commence listening and dispatching of events.
+func (s *source) Start(handler resource.EventHandler) error {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	if s.stopCh != nil {
+		return fmt.Errorf("already synchronizing resources: name='%s', gv='%v'",
+			s.GetSpec().Singular, s.GetSpec().GroupVersion())
+	}
+	if handler == nil {
+		return fmt.Errorf("invalid handler")
+	}
+
+	log.Scope.Debugf("Starting source for %s(%v)", s.GetSpec().Singular, s.GetSpec().GroupVersion())
+
+	s.stopCh = make(chan struct{})
+	s.handler = handler
+
+	s.informer = s.NewInformer(s.sharedInformers)
```

Comment:
This was confusing to read. It looks like this is using the *Processor, right? Can we make it explicit (i.e. give it a full field name such as "processor"?)